File: IOS/TestEyes/TestEyes/GADMasterViewController.m
//
//  GADMasterViewController.m
//  100Numbers
//
//  Created by Binh Du  on 4/12/15.
//  Copyright (c) 2015 LapTrinhAlgo.Com. All rights reserved.
//

#import "GADMasterViewController.h"

@implementation GADMasterViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



+(GADMasterViewController *)GetSingleton
{
    static dispatch_once_t pred;
    static GADMasterViewController *shared;
    
    // Will only be run once, the first time this is called
    dispatch_once(&pred, ^
                  {
                      NSLog(@"GADMaster:GetSingeton");
                      shared = [[GADMasterViewController alloc] init];
                  });
    return shared;
}


-(id)init
{
    NSLog(@"GADMaster:Init");
    if (self = [super init])
    {
        // Has an ad request already been made
        m_Banner = [[GADBannerView alloc] init];
        m_IsLoaded = NO;
        
        [self GetInterstitialAds];
    }
    
    
    return self;
}


-(void)resetAdBannerView:(UIViewController *)p_RootViewController AtFrame:(CGRect) frm
{
    NSLog(@"GADMaster:resetAdBannerView with frame:%f-%f", frm.size.width, frm.size.height);
    
    m_Banner.frame = frm;
    
    // Ad already requested, simply add it into the view
    if (m_IsLoaded)
    {
        [p_RootViewController.view addSubview:m_Banner];
    }
    else
    {
        m_Banner.delegate = self;
        m_Banner.rootViewController = p_RootViewController;
        m_Banner.adUnitID = AMOD_BANNER_FOOTER_UNIT;
        
        GADRequest *request = [GADRequest request];
        //request.testDevices = @[ @"15b5d334e2e980a8595f943d3fe28621" ];
        [m_Banner loadRequest:request];
        [p_RootViewController.view addSubview:m_Banner];
        //m_IsLoaded = YES;
    }
}

- (void)GetInterstitialAds
{
    NSLog(@"GADMaster:GetInterstitialAds");
    m_Interstitial = [[GADInterstitial alloc] initWithAdUnitID:AMOD_INTERSTITIAL_UNIT];
    m_Interstitial.delegate = self;
    GADRequest *request = [GADRequest request];
    //request.testDevices = @[ @"15b5d334e2e980a8595f943d3fe28621" ];
    [m_Interstitial loadRequest:request];
}

-(void)ResetAdInterstitialView:(UIViewController *)rootViewController
{
    NSLog(@"GADMaster:ResetAdInterstitialView");
    if([m_Interstitial hasBeenUsed])
    {
        NSLog(@"GADMaster:Interstitial has been used");
        //return;
    }
    else if([m_Interstitial isReady])
    {
        [m_Interstitial presentFromRootViewController:rootViewController];
    }
    else
    {
        NSLog(@"GADInterstitial not ready");
        [self GetInterstitialAds];
    }
    
}

#pragma GADDelegate
- (void)interstitialDidDismissScreen:(GADInterstitial *)p_interstitial
{
    NSLog(@"GADMaster:DidDismissinterstitial");
    [self GetInterstitialAds];
}

- (void)adViewDidReceiveAd:(GADBannerView *)adView
{
    NSLog(@"GADMaster:adViewDidReceiveAd");
    
    adView.alpha = 0;
    [UIView animateWithDuration:1.0 animations:^{
        adView.alpha = 1;
    }];
    
    m_IsLoaded = TRUE;
    
}

- (void) adViewWillDismissScreen:(GADBannerView *)bannerView
{
    NSLog(@"GADMaster:adViewWillDismissScreen");
    m_IsLoaded = FALSE;
}

- (void) adView:(GADBannerView *)bannerView didFailToReceiveAdWithError:(GADRequestError *)error
{
    NSLog(@"Failed to receive ad with error: %@", [error localizedFailureReason]);
    m_IsLoaded = FALSE;
}

- (void) adViewWillPresentScreen:(GADBannerView *)bannerView
{
    NSLog(@"adViewWillPresentScreen");
}

@end

File: IOS/Funny Stories/Funny Stories/GADMasterViewController.swift
//
//  GADMasterViewController.swift
//  Funny Stories
//
//  Created by Nguyễn Thế Bình on 12/23/15.
//  Copyright © 2015 Nguyễn Thế Bình. All rights reserved.
//

import UIKit
import GoogleMobileAds

class GADMasterViewController: NSObject, GADBannerViewDelegate, GADInterstitialDelegate
{
    static let ShareInstance = GADMasterViewController()
    
    var m_Banner: GADBannerView
    var m_Interstitial: GADInterstitial;
    var m_IsLoaded: Bool;
    
    
    private override init()
    {
        NSLog("GADMaster:Init")
        m_Banner = GADBannerView()
        m_IsLoaded=false
        m_Interstitial = GADInterstitial()
        
        super.init()
        self.GETInterstitialAds()
    }

    
    func ResetBannerView(p_view: UIViewController?, p_frame: CGRect) -> Bool
    {
        if p_view != nil
        {
            print("GADMaster:resetAdBannerView with frame:%f-%f",p_frame.size.width,p_frame.size.height)
            m_Banner.frame = p_frame
            if m_IsLoaded
            {
                m_Banner.rootViewController = p_view
                p_view!.view.addSubview(m_Banner)
            }
            else
            {
                m_Banner.delegate   =   self
                m_Banner.rootViewController = p_view
                m_Banner.adUnitID   =   AMOD_BANNER_FOOTER_UNIT
                m_Banner.loadRequest(GADRequest())
                p_view!.view.addSubview(m_Banner)
            }
            
            return true
        }
        else
        {
            return false
        }

    }
    
    
    func GETInterstitialAds()
    {
        NSLog("GADMaster:GetInterstitialAds")
        m_Interstitial = GADInterstitial(adUnitID:AMOD_INTERSTITIAL_UNIT)
        m_Interstitial.delegate = self
        m_Interstitial.loadRequest(GADRequest())
    }
    
    
    func ResetInterstitialView(p_view: UIViewController)
    {
        NSLog("GADMaster:ResetAdInterstitialView")
        if m_Interstitial.hasBeenUsed
        {
            NSLog("GADMaster:Interstitial has been used")
            
        }
        else
        {
            if m_Interstitial.isReady
            {
                m_Interstitial.presentFromRootViewController(p_view)
                
            }
            else
            {
                NSLog("GADInterstitial not ready")
                self.GETInterstitialAds()
                
                sleep(2)
                if m_Interstitial.isReady
                {
                    m_Interstitial.presentFromRootViewController(p_view)
                }
                
            }
            
        }
        
    }
    
    //
    // GADelegate
    //
    
    func interstitialWillDismissScreen(ad: GADInterstitial!)
    {
        print("GADMaster:WillDismissScreen")
    }
    
    func interstitialDidDismissScreen(ad: GADInterstitial!)
    {
        print("GADMaster:DidDismissinterstitial")
        self.GETInterstitialAds()
    }
    
    
    func adViewDidReceiveAd(bannerView: GADBannerView!)
    {
   
        NSLog("GADMaster:adViewDidReceiveAd")
        bannerView.alpha=0
        UIView.animateWithDuration(1, animations:
        {
            bannerView.alpha=1
        })
        
        m_IsLoaded = true
    }
    
    func adViewWillDismissScreen(bannerView: GADBannerView!)
    {
         NSLog("GADMaster:adViewWillDismissScreen")
    }

    func adViewDidDismissScreen(bannerView: GADBannerView!)
    {
        NSLog("GADMaster:adViewDidDismissScreen")
        m_IsLoaded = false
    }
    
    func adViewWillPresentScreen(bannerView: GADBannerView!)
    {
        NSLog("adViewWillPresentScreen")
    }
 
    
    func adView(bannerView: GADBannerView!, didFailToReceiveAdWithError error: GADRequestError!)
    {
        NSLog(error.localizedFailureReason!)
        print("Rootview: ", m_Banner.rootViewController)
        
        m_IsLoaded = false
        
        //bannerView.loadRequest(GADRequest())
    }
}
